wpiutil/src/main/native/include/wpi/telemetry/TelemetryTable.h

```diff
@@ -0,0 +1,346 @@
+   *
+   * @param path path with trailing "/".
+   */
+  explicit TelemetryTable(std::string_view path, const private_init&)
+      : m_path{path} {}
+
+  TelemetryTable(const TelemetryTable&) = delete;
+  TelemetryTable& operator=(const TelemetryTable&) = delete;
+
+  /**
+   * Gets the table path.
+   *
+   * @return path with trailing "/"
+   */
+  std::string_view GetPath() const;
+
+  /**
+   * Gets a child telemetry table.
+   *
+   * @param name table name
+   * @return table
+   */
+  TelemetryTable& GetTable(std::string_view name) const;
+
+  /**
+   * Indicates duplicate values should be preserved. Normally duplicate values
+   * are ignored.
+   *
+   * @param name the name
+   */
+  void KeepDuplicates(std::string_view name);
+
+  /**
+   * Sets property for a value. Properties are stored as a key/value map.
+   *
+   * @param name the name
+   * @param key property key
+   * @param value property value
+   */
+  void SetProperty(std::string_view name, std::string_view key,
+                   std::string_view value);
+
+  /**
+   * Sets custom data type for a value. Generally not necessary.
+   *
+   * @param name the name
+   * @param typeString type string
+   */
+  void SetTypeString(std::string_view name, std::string_view typeString);
+
+  /**
+   * Logs an object.
+   *
+   * @param name the name
+   * @param value the value
+   * @param info type parameters for struct serializer (optional)
+   */
+  template <typename T, typename... I>
+  void Log(std::string_view name, const T& value, I... info) {
+    if constexpr (impl::TelemetryTableLoggableADL<T, I...>) {
+      impl::TelemetryLogTableADL(GetTable(name), value, info...);
+    } else if constexpr (impl::TelemetryEntryLoggableADL<T, I...>) {
+      impl::TelemetryLogEntryADL(*this, name, value, info...);
+    } else if constexpr (wpi::StructSerializable<T, I...>) {
+      using S = wpi::Struct<T, I...>;
+      TelemetryRegistry::AddStructSchema<T>(info...);
+      if constexpr (sizeof...(I) == 0) {
+        if constexpr (wpi::is_constexpr([] { S::GetSize(); })) {
+          uint8_t buf[S::GetSize()];
+          S::Pack(buf, value);
+          LogRaw(name, S::GetTypeString(), buf);
+          return;
+        }
+      }
+      wpi::SmallVector<uint8_t, 128> buf;
+      buf.resize_for_overwrite(S::GetSize(info...));
+      S::Pack(buf, value, info...);
+      LogRaw(name, S::GetTypeString(info...), buf);
+    } else if constexpr (wpi::ProtobufSerializable<T>) {
+      wpi::ProtobufMessage<T> msg;
+      TelemetryRegistry::AddProtobufSchema<T>(msg);
+      wpi::SmallVector<uint8_t, 128> buf;
+      msg.Pack(buf, value);
+      LogRaw(name, msg.GetTypeString(), buf);
+    } else if constexpr (std::is_integral_v<T>) {
+      Log(name, static_cast<int64_t>(value));
+    } else if constexpr (std::is_floating_point_v<T>) {
+      Log(name, static_cast<double>(value));
+    } else {
+      // TODO: see if it's convertable with fmt::to_string?
```

Comment:
Based on https://en.cppreference.com/w/cpp/language/static_assert, pre C++26 the message must be a string literal.